src/common/ad_mem.c

```diff
@@ -3,5 +3,1857 @@
+	bd->bd_size		= blob_size(blob);
+	bd->bd_arena_size	= ARENA_SIZE;
+
+	/* register the default arena */
+	rc = blob_register_arena(blob, 0, grp_specs_def, ARRAY_SIZE(grp_specs_def), NULL);
+	D_ASSERT(rc == 0); /* no reason to fail */
+
+	/* create arena 0 (ad_blob_df is stored in the first 32K of it) */
+	rc = arena_reserve(blob, 0, NULL, &arena);
+	D_ASSERT(rc == 0);
+	D_ASSERT(arena->ar_df);
+	D_ASSERT(arena->ar_df->ad_id == 0);
+
+	/* NB: no transaction, write arena[0] and super block to storage straight away */
+	rc = arena_tx_publish(arena, NULL);
+	if (rc)
+		goto failed;
+
+	arena->ar_unpub = false;
+
+	blob->bb_arena_last[0]	= bd->bd_asp[0].as_arena_last;
+	/* already published arena[0], clear the reserved bit */
+	clrbit(blob->bb_bmap_rsv, arena->ar_df->ad_id);
+
+	ad_iod_set(&iod, blob_ptr2addr(blob, arena->ar_df), ARENA_HDR_SIZE + BLOB_HDR_SIZE);
+	ad_sgl_set(&sgl, &iov, arena->ar_df, ARENA_HDR_SIZE + BLOB_HDR_SIZE);
+
+	D_ASSERT(store->stor_ops);
+	rc = store->stor_ops->so_write(store, &iod, &sgl);
+	if (rc) {
+		D_ERROR("Failed to write ad_mem superblock\n");
+		goto failed;
+	}
+	arena_decref(arena);
+
+	D_DEBUG(DB_TRACE, "Ad-hoc memory blob created\n");
+	blob_set_opened(blob);
+	return 0;
+failed:
+	if (arena)
+		arena_decref(arena);
+	blob_handle_release(bh);
+	return rc;
+}
+
+int
+ad_blob_prep_open(char *path, struct ad_blob_handle *bh)
+{
+	struct ad_blob	*blob;
+	bool		 is_dummy = false;
+
+	if (!strcmp(path, DUMMY_BLOB))
+		is_dummy = true;
+
+	if (!is_dummy) {
+		/* XXX dummy is the only thing can be supported now */
+		D_ASSERT(0);
+	} else {
+		if (dummy_blob) {
+			blob = dummy_blob;
+			blob_addref(blob);
+		} else {
+			D_ALLOC_PTR(blob);
+			if (!blob)
+				return -DER_NOMEM;
+
+			blob->bb_fd = -1;
+			blob->bb_ref = 1;
+			blob->bb_dummy = true;
+		}
+	}
+	bh->bh_blob = blob;
+	return 0;
+}
+
+int
+ad_blob_post_open(struct ad_blob_handle bh)
+{
+	struct ad_blob		*blob  = bh.bh_blob;
+	struct umem_store	*store = &blob->bb_store;
+	struct ad_blob_df	*bd;
+	struct umem_store_iod	 iod;
+	d_sg_list_t		 sgl;
+	d_iov_t			 iov;
+	int			 rc;
+
+	if (blob->bb_opened)
+		return 0;
+
+	D_ALLOC_PTR(bd);
+	if (!bd)
+		return -DER_NOMEM;
+
+	/* blob header is stored right after header of arena[0] */
+	ad_iod_set(&iod, blob->bb_store.stor_addr + ARENA_HDR_SIZE, sizeof(*bd));
+	ad_sgl_set(&sgl, &iov, bd, sizeof(*bd));
+
+	/* read super block to temporary buffer */
+	D_ASSERT(store->stor_ops);
+	rc = store->stor_ops->so_read(store, &iod, &sgl);
+	if (rc) {
+		D_ERROR("Failed to read superblock of ad_mem\n");
+		goto failed;
+	}
+
+	if (bd->bd_magic != BLOB_MAGIC || bd->bd_version == 0) {
+		D_ERROR("Invalid superblock: magic=%x, version=%d\n",
+			bd->bd_magic, bd->bd_version);
+		goto failed;
+	}
+
+	store->stor_size = bd->bd_size;
+	store->stor_addr = bd->bd_addr;
+	blob->bb_pgs_nr = ((blob_size(blob) + ARENA_SIZE_MASK) >> ARENA_SIZE_BITS);
+
+	rc = blob_init(blob);
+	if (rc)
+		goto failed;
+
+	rc = blob_load(blob);
```

Comment:
I think the whole tmpfs loading (including the ad superblock) should be moved to generic umem interface and being shared by BMEM, allocator can assume the data is already loaded from SSD to tmpfs.

Reply:
the thing bothered me is, page cache is supposed to be managed by the allocator... we have to allocate pages while loading information from SSD right?

Reply:
I agree, I actually thought about this, but I decided to put them together in the first step, because it is little difficult to abstract things before I have a few pieces. I can do this in follow-on patches.